### ISSUE TITLE
tag --tag-all needs behavior clarity

### ISSUE THREAD

**Issue by shortjared** [jaws-framework/JAWS#164]
`jaws tag --tag-all` expects lambda or endpoint to be defined.

However, I was confused because I ran `jaws tag --tag-all` and it appeared to default to tagging only lambdas. See https://github.com/jaws-framework/JAWS/blob/b52b338f209fd38c3799d2cc26fae183d2f44f6e/bin/jaws#L174

```javascript
  .action(function(type, options) {
      type = type || 'lambda';
```

It took me looking at the source to understand what was going on.

I would expect either...
a) Explicit failure with error that says an option is required
b) Everything (both lambdas and endpoints) are tagged

I'd be happy to jump in and contribute, just need to know if you want me to take action and which option is preferred.

**Comment by boushley** [comment]
Good catch!

I would personally vote for option b, tag all without a type flag tags everything.

We should probably get an opinion from @doapp-ryanp or @ac360 ?

**Comment by doapp-ryanp** [comment]
I can see this has already been merged, however I vote for a).  

The `type` (see jaws tag -h) is both misleading and currently incorrect IMO.  I think `type` should be changed from optional (`[type]`) to required (`<type>`) and I think the type of `all` should be changed to `both` to be more clear.  I don't vote b) because anytime you do a wide spread modification of something I like to be explicit AND I think 90% of the tag's are going to be to lambdas (since once you setup an APIG endpoint, you prob wont change it much, but you will be many lambda updates throught the lifetime of the lambda).

**Comment by shortjared** [comment]
That's reasonable. I usually favor explicit but to me `--tag-all` felt like an explicit "all" with optional specification. I can fix that and do another PR.

**Comment by boushley** [comment]
Ryan I see what you mean about frequency of change, but is that what this
tagging concept is about? The tragging idea feels more like is this prod
ready or not. You can deploy just your lambdas easily via `jaws deploy ` so
tagging everything doesn't really change how things are deployed.

If tagging is about what has changed since the last time you deployed we
may want to make that a bit more clear probably with a name change, auto
untag on deploy, and really you would want to tag per stage.

I like the idea of a tag as whether this belongs in prod or is under
development, and in that case the lambda gateway change frequency
distinction is less important.

**Comment by doapp-ryanp** [comment]
@boushley the `jaws tag` command is intended to save time when deploying, has nothing to do with the concept of stage promotion.

Lets say, for example, you have 10 lambdas.  You change some common code that impacts all of them.  Instead of having to cd into each of the 10 lambda dirs and running `jaws deploy lambda`, you can tag all of them and then run `jaws deploy lambda <stage> <region> -t` which will deploy all lambdas "tag"ged as needing deployment - in this case all lambdas.

I totally agree a name change is in order, since it is def confusing.  Got any ideas?

**Comment by joostfarla** [comment]
What do you think about `jaws prepare`?

**Comment by shortjared** [comment]
Does it make sense to use something like `jaws prepare` as mentioned, in combination with arbitrary tagging which could be useful for something like state promotion? Perhaps that is getting to much into state  / process management.

In awsm-module...
`jaws tag <tag>`

`jaws prepare <tag>`
`jaws deploy lambda <stage> <region> --prepared`

**Comment by boushley** [comment]
Stage promotion really does feel like a separate feature, and it would be
best implemented by AWS. So let's not try to tackle that here.

So then there's the question of what tag should do. I wouldn't use it,
except that it's the easiest way to deploy everything in a CI setup. So for
my team this has taken on the semantic meaning of whether something is
ready to be deployed, or still under development and should only be
manually deployed for testing. I think this makes sense with `deploy: true
`

Prepping for a singular deployment with this command feels odd, especially
since the awsm.json file where this is marked is likely in source control.
Preparing for a single deploy feels like an ephemeral task.

Then you could make the argument for real tagging that lets you have named
tags associated with items in a many to many relationship. This would be a
significant amount of piping throughout the library though, although
probably the best long term feature.

That said I don't think this is in any way among the most pressing issues
for the framework at this time. There are many other features I intend to
work on first.

Aaron

**Comment by shortjared** [comment]
So perhaps the best approach is to let the feature stand as is, and circle back around when other things are more stable. Particularly after dependency management is sorted.

`jaws tag --[un]tag-all lambda|endpoint|both` works, it just isn't pretty and doesn't serve a whole lot of purpose, especially if you are using the deploy tagging meaningfully in a CI process.

**Comment by doapp-ryanp** [comment]
yep agree on letting it stand as is, we will circle back. lots bigger features to work on ATM.

**Comment by fabiopelosin** [comment]
One thing that is not clear to me is why the tag is saved in the `asm.json` file. I would consider a more unix like behaviour to pass the paths of the modules to deploy. For example:

```console
$ jaws deploy 'back/aws_modules/users/get' 'back/aws_modules/users/post'
```

This approach would have the following benefits:

- Allow to use the shell history for deployments. In my workflow sometimes I end up troubleshooting a deployment. Although I might end up redeploying numerous times this task is ephemeral. Shell history is extremely convenient for this behaviour.
- Avoid storing the deployment information in the SCM history. In my opinion storing the `"deploy": true` flag is similar to storing files being edited in the editor, and thus this information doesn't belong in the SCM. 
- Script automation.
- Allow to deploy in one command without requiring to go trough the `jaws tag` command.

**Comment by doapp-ryanp** [comment]
@fabiopelosin love it. Reason its not that way was lack of time :)

**Comment by doapp-ryanp** [comment]
PR would be cool here. make sure to leave the default of no path specified deploys CWD

**Comment by shortjared** [comment]
I believe this actually renders tagging in its current form redundant at best right? Should it be pulled to avoid confusion later on?

**Comment by doapp-ryanp** [comment]
@shortjared agreed it makes it redundant - so we are killing 2 birds with one stone here (as discussed above the tag name was confusing in its self).

FWIW, we are trying to change as little as possible core function in the mainline in prep for our AWS presentation next week.  So feel free to send PR, just don't get feelings hurt if I dont merge right away :)

**Comment by shortjared** [comment]
I won't have time to get to this for a couple days so if anyone else wants to jump in and grab it feel free.

@doapp-ryanp Totally understand, looking forward to the presentation!

**Comment by doapp-ryanp** [comment]
I have just added this to the WIP es6-modular branch.  It now looks like the other unix cmds ur familiar with. Even added a handy `ALL` keyword :)

`$ jaws lambda deploy [options]... [paths|ALL]...`

See https://github.com/jaws-framework/JAWS/commit/558e784c35b07f767c2442864c7b795e00b7993a
